---
title: "Gibbs_Sampling Optimization"
author: "Marcus Tutert"
date: "04/10/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#Build the package (always good to update!)
system("R CMD INSTALL ~/Desktop/Oxford_Dphil/InferLD")
library(InferLD)
```
First things first, let us go and check that our packages are running fine
```{r testing_packages, echo=FALSE}
library(InferLD)
#Set number of haplotypes and variants to a small number so we can do testing and debugging
nhaps = 100 #Haplotypes in the reference panel
nsnps = 100 #Variants in the reference panel

## path = dirname(getwd()) #Set path relative to testing folder
path <- getwd()

#Load in 1000G data from ./Test/Data
ref_hap_panel     = as.matrix(read.table(sprintf('%s/Data/ref_panel_filtered',path),
                                         header = TRUE))

#Subset to nhaps and nsnps size to make testing easier
ref_hap_panel     = ref_hap_panel[ 1:nhaps, 1:nsnps ]

#Get AF of the variants in the reference pnale
ref_variants_af   = colMeans(ref_hap_panel)
Fst               = 1e-2

#Assume that allele frequencies are the same in the reference panel and the GWAS panel
sample_weights           = rgamma(n = nhaps, shape =  1 / ( nhaps * (Fst/(1-Fst))), scale =  1 / ( nhaps * (Fst/(1-Fst))))
normalize_sample_weights = sample_weights/(sum(sample_weights))
gwas_variants_af         = colSums(ref_hap_panel * normalize_sample_weights)

#Filter out SNPs that are above/below a threshhold (0-100% exclusive)
filter_snp_index        = which(colMeans(ref_hap_panel) > 0.99)
filter_snp_index_2      = which(colMeans(ref_hap_panel) < 0.01)
filter_snp_index_merged = c(filter_snp_index,filter_snp_index_2)
if (length(filter_snp_index_merged)>0) {
  ref_hap_panel     = ref_hap_panel[,-filter_snp_index_merged]
  gwas_variants_af  = gwas_variants_af[-filter_snp_index_merged]
  ref_variants_af   = ref_variants_af[-filter_snp_index_merged]
}

#Transform AF's to SE
nSamples          = 100
alpha             = 1e3
#fig
observed_gwas_se  = 1/(gwas_variants_af*(1-gwas_variants_af))
test           = LD_from_GSHMM(ref_panel_haplotypes = ref_hap_panel,
                                   Fst                  = 1e-1,
                                   betas                = FALSE,
                                   alpha                = alpha,
                                   nSamples             = nSamples,
                                   recomb_rate          = 1e-10,
                                   weights_resolution   = 5,
                                   likelihood           = TRUE,
                                   se_observed          = observed_gwas_se,
                                   LD_Infer             = FALSE,
                                   genetic_map          = FALSE,
                                   chain_likelihood     = TRUE,
                                   nChains              = 2)

```
