#For UKBB simulations, everything will (largely) stay the same, except for the input data
#Stick with my QCd genotyped WBA cohort (BNeale) to perform the data-analysis
#As reference panels, I will use the different super-pops from 1000G analysis
#Can do this across multiple regions, all simultaneously on SNAKEMAKE

# #Start with 1000G haplotypes (Chr21) data and extract the different super population sample labels
# super_pops = c("SAS","EUR","AMR","AFR","EAS")
# samples = read.table("/well/mcvean/mtutert/1000_Genomes/VCF/integrated_call_samples_v3.20130502.ALL.panel", header = T)
# #Read in sample file
# for(i in super_pops){
#   print(i)
#   filtered_pop = samples[samples$super_pop == i,1]
#   #Write out population regions
#   write.table(filtered_pop,sprintf("%s_samples", i), quote = F, col.names = F, row.names = F)
# }
#
# #Now we want to break up the rsids into 500 chunks each
# regions    = 10
# nsnps      = 500 ####Note this is before any filtering
# for (i in super_pops) {
#   #Loop through the super pops
#   for (j in 1:regions) {
#     #Get region stepsize
#     stepsize = seq(1,nsnps*regions+1,nsnps)
#     #Loop through panels
#     start_SNP = stepsize[j]
#     end_SNP   = stepsize[j+1]-1
#     #Read in the SNPs rsid from the file R (generated by the freq command)
#     snp_ids   = read.table("/well/mcvean/mtutert/1000_Genomes/prior_draw_analysis_panels/all_snp_rsid.afreq", header = T)
#     write.table(snp_ids[start_SNP:end_SNP,2],"snp_rsid", quote = F, row.names = F, col.names = F)
#     system(sprintf("/well/mcvean/mtutert/software/plink2 --vcf ALL.chr22.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz --extract snp_rsid --out 1000G_%s_region_%s --export haps --max-alleles 2",i,j))
#   }
# }

############
#Let us assume for now our snakemake params are the super population (EUR/AMR...)
#And we know what region chunk we are extract (SNPs 1:500, 501;1000...)
#Might need to do some string splitting etc here
