---
title: "Gibbs_Sampling_Optimzation"
author: "Marcus Tutert"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document:
    
    toc: yes
    toc_depth: 3
    number_sections: yes
    theme: united
    highlight: tango
    css: my.css
    always_allow_html: true
  pdf_document:
    toc: yes
    toc_depth: '3'
---
```{r setup, include=FALSE}
#Check if all packages work!
#Build the package (always good to update!)
pkg = "/Users/marcustutert/Desktop/Oxford_Dphil/InferLD"
devtools::build(pkg = pkg)
devtools::document(pkg)
devtools::load_all(pkg, quiet = FALSE)
package_tarball <- devtools::build(pkg = pkg, manual = TRUE)
install.packages(package_tarball, repos = NULL, type ="source")
library(InferLD)
library(plotly)
```
Look at effect of changing nHaps with some drift
```{r nHaps Function, echo=FALSE}
#Loop over a given number of replicates (here it will be 5)
nreplicates         = 10
replicate_result    = list()  
count               = 1   #Bookmarking thru the double loop
gwas_variants_af    = list()                  #Going to store this as a list!

for (rep_number in 1:nreplicates) {
  print(sprintf("We are on replicate %s", rep_number))
  #Parameters
  nhaps               = c(10,100,400,800) #Change the Nhaps in reference panel
  nsnps               = 100                      #Constant NSnps
  results             = c()                     #Store results as a vector
  Fst                 = 1e-1                    #Constant Fst (High Genetic Drift)
  nSamples            = 5                       #Constant nSamples
  alpha               = 1e3                     #Constant alpha
  weights_resolution  = 5
  for (i in 1:length(nhaps)) {
    chains                = 1                   #Single Chain
    sim_data              = simulate_genetic_drift(Fst   = Fst, #Same Fst as before
                                                   nhaps = nhaps[i],
                                                   nsnps = nsnps, 
                                                   noise = alpha,
                                                   weights_resolution = 200)
    observed_gwas_se                    = sim_data[[1]]
    gwas_variants_af[[count]]           = sim_data[[2]]  #Store the GWAS variant AF as a list
    ref_hap_panel                       = sim_data[[3]]         
    results[i]                          = LD_from_GSHMM(ref_panel_haplotypes  = ref_hap_panel,
                                                         Fst                  = Fst,
                                                         betas                = FALSE,
                                                         alpha                = alpha,
                                                         nSamples             = nSamples,
                                                         recomb_rate          = 1e-100,
                                                         weights_resolution   = weights_resolution,
                                                         likelihood_toggle    = TRUE,
                                                         se_observed          = observed_gwas_se,
                                                         LD_Infer             = FALSE,
                                                         genetic_map          = FALSE,
                                                         chain_likelihood     = TRUE,
                                                         nChains              = chains,
                                                         recombination        = FALSE)
  count = count +1 #Update bookmarks
  }
  replicate_result[[rep_number]] = results
}

#Add on the traces across the different parameters (ie nHaps)
p <- plot_ly()
col     = brewer.pal(n = length(nhaps), name = "Dark2")
count = 1
r2    = list()
for (rep_number in 1:nreplicates) {
  for (i in 1:length(nhaps)) {
  ll = replicate_result[[rep_number]][i][[1]][[3]]
  nSamples_trace = seq(1:length(ll))/nhaps[i]
  r2[[count]] = r2_true_inferred_af_across_chains(inferred_allele_freq = replicate_result[[rep_number]][i][[1]][[4]],
                                         true_allele_freq     = gwas_variants_af[[count]],
                                         graphic              = FALSE,
                                         nSamples             = nSamples,
                                         nhaps                = nhaps[i])
  if (length(nSamples_trace) != length(r2)) {
    
  }

  if (rep_number == 1 ) {
     p <- add_trace(p, x=nSamples_trace, y=r2[[count]], type="scatter", mode="lines",evaluate = TRUE, 
                    name = sprintf("%s Ref Haplotypes",nhaps[i]),color=col[i], showlegend = T, opacity = 0.2)
    }
  else{
    p <- add_trace(p, x=nSamples_trace, y=r2[[count]], type="scatter", mode="lines",evaluate = TRUE, 
                    name = sprintf("%s Ref Haplotypes",nhaps[i]),color=col[i], showlegend = F, opacity = 0.2)
    }
    count = count + 1
  }
}

#Average across the r2's
#Split list into the groups
groups = list()
for (i in 1:length(nhaps)) {
  groups[[i]] = r2[seq(i,40,4)]
}

#Average 
r2_avg = list()
for (i in 1:length(groups)) {
    reshape = length(replicate_result[[1]][i][[1]][[3]])
    avg = rowMeans(matrix(unlist(groups[[i]]),reshape))
    r2_avg[[i]] = avg
    nSamples_trace = length(unlist(r2_avg[[i]]))
}

for (idx in 1:length(nhaps)) {
      ll = replicate_result[[rep_number]][idx][[1]][[3]]
      
      nSamples_trace = seq(1:length(ll))/nhaps[idx]
      p <- add_trace(p, x= nSamples_trace, y=unlist(r2_avg[[idx]]), 
                    type="scatter", mode="lines",evaluate = TRUE, 
                    showlegend = F, opacity = 1,color = col[idx])
}

p <- p %>% layout(title = 'Correlation of True to Inferred Allele Frequency Across Sweeps',
                    xaxis = list(title = 'nSamples (Full Sweeps)'),
                    yaxis = list(title = 'r2 correlation (Inferred~True)'))
p

ll_plot   = plot_ly(showlegend = TRUE)
ll_zoomed = plot_ly(showlegend = FALSE)
max_ll  = c()
col     = brewer.pal(n = length(nhaps), name = "Dark2")
for (i in 1:length(nhaps)) {
  ll = results[i][[1]][[3]]
  max_ll[i]  = max(ll)
  ll_plot <- add_trace(ll_plot,
                       x=seq(1:length(ll))/nhaps[i], 
                       y=ll, type="scatter", 
                       mode="markers",
                       evaluate = TRUE, name = sprintf("%s Ref Haplotypes",nhaps[i]),
                       marker   = list(color = col[i]))
  ll_zoomed <- add_trace(ll_zoomed,
                       x=seq(1:length(ll))/nhaps[i], 
                       y=ll, type="scatter", 
                       mode="markers",
                       evaluate = TRUE, name = sprintf("%s Ref Haplotypes",nhaps[i]),
                       marker   = list(color = col[i]))
}
ll_plot   = ll_plot %>% layout(title = 'Log Likelihood Across per Haplotype Sweeps',
                    xaxis = list(title = 'nSamples (Per Haplotype Sweeps)'),
                    yaxis = list(title = 'Log Likelihood'))
ll_zoomed = ll_zoomed  %>% layout(title = 'Log Likelihood Across per Haplotype Sweeps',
                    xaxis = list(title = 'nSamples (Per Haplotype Sweeps)'),
                    yaxis = list(range = c(0,max(max_ll)*1.1),
                                 title = 'Log Likelihood'))
ll_zoomed <- ll_zoomed %>% layout(marker     = list(color = col))

fig <- subplot(ll_plot,
               ll_zoomed) %>%  add_annotations(
        text = "Number of Samples (Per Haplotype Reference Panel Update)",
        x = 0.5,
        y = 0,
        yref = "paper",
        xref = "paper",
        xanchor = "center",
        yanchor = "bottom",
        yshift = -35,
        showarrow = FALSE,
        font = list(size = 15) 
    ) 
fig %>% layout(yaxis = list(title = "Log-Likelihood"))
```
Look at effect of changing nHaps WITH drift
```{r}
nhaps               = c(10,20,50,100,400,800) #Change the Nhaps in reference panel
nsnps               = 50                   #Constant NSnps
results             = c()                   #Store results as a vector
gwas_variants_af    = list()        #Needs to be a list 
Fst                 = 1e-1
nSamples            = 5      #Constant nSamples
alpha               = 1e3     #Constant alpha
weights_resolution  = weights_resolution
for (i in 1:length(nhaps)) {
  chains                = 1       #Single Chain
  sim_data              = simulate_genetic_drift(Fst   = Fst, 
                                                 nhaps = nhaps[i],
                                                 nsnps = nsnps, 
                                                 noise = alpha,
                                                 weights_resolution = 200)
  observed_gwas_se      = sim_data[[1]]
  gwas_variants_af[[i]] = sim_data[[2]]  #Store the GWAS variant AF as matrix 
  ref_hap_panel         = sim_data[[3]]         
  results[i]            = LD_from_GSHMM(ref_panel_haplotypes = ref_hap_panel,
                                         Fst                  = Fst,
                                         betas                = FALSE,
                                         alpha                = alpha,
                                         nSamples             = nSamples,
                                         recomb_rate          = 1e-100,
                                         weights_resolution   = 200,
                                         likelihood_toggle    = TRUE,
                                         se_observed          = observed_gwas_se,
                                         LD_Infer             = FALSE,
                                         genetic_map          = FALSE,
                                         chain_likelihood     = TRUE,
                                         nChains              = chains,
                                         recombination        = FALSE)
}

#Add on the traces across the different parameters (ie nHaps)
p <- plot_ly()
for (i in 1:length(nhaps)) {
  ll = results[i][[1]][[3]]
  nSamples_trace = seq(1:length(ll))/nhaps[i]
  r2 = r2_true_inferred_af_across_chains(inferred_allele_freq = results[i][[1]][[4]],
                                         true_allele_freq     = gwas_variants_af[[i]],
                                         graphic              = FALSE,
                                         nSamples             = nSamples,
                                         nhaps                = nhaps[i])
  p <- add_trace(p, x=nSamples_trace, y=r2, type="scatter", mode="markers",evaluate = TRUE, name = sprintf("%s Ref Haplotypes",nhaps[i]))
}
p <- p %>% layout(title = 'Correlation of True to Inferred Allele Frequency Across Sweeps',
                    xaxis = list(title = 'nSamples (Full Sweeps)'),
                    yaxis = list(title = 'r2 correlation (Inferred~True)'))
p

ll_plot   = plot_ly(showlegend = TRUE)
ll_zoomed = plot_ly(showlegend = FALSE)
max_ll  = c()
col     = brewer.pal(n = length(nhaps), name = "Dark2")
for (i in 1:length(nhaps)) {
  ll = results[i][[1]][[3]]
  max_ll[i]  = max(ll)
  ll_plot <- add_trace(ll_plot,
                       x=seq(1:length(ll))/nhaps[i], 
                       y=ll, type="scatter", 
                       mode="markers",
                       evaluate = TRUE, name = sprintf("%s Ref Haplotypes",nhaps[i]),
                       marker   = list(color = col[i]))
  ll_zoomed <- add_trace(ll_zoomed,
                       x=seq(1:length(ll))/nhaps[i], 
                       y=ll, type="scatter", 
                       mode="markers",
                       evaluate = TRUE, name = sprintf("%s Ref Haplotypes",nhaps[i]),
                       marker   = list(color = col[i]))
}
ll_plot   = ll_plot %>% layout(title = 'Log Likelihood Across per Haplotype Sweeps',
                    xaxis = list(title = 'nSamples (Per Haplotype Sweeps)'),
                    yaxis = list(title = 'Log Likelihood'))
ll_zoomed = ll_zoomed  %>% layout(title = 'Log Likelihood Across per Haplotype Sweeps',
                    xaxis = list(title = 'nSamples (Per Haplotype Sweeps)'),
                    yaxis = list(range = c(0,max(max_ll)*1.1),
                                 title = 'Log Likelihood'))
ll_zoomed <- ll_zoomed %>% layout(marker     = list(color = col))

fig <- subplot(ll_plot,
               ll_zoomed) %>%  add_annotations(
        text = "Number of Samples (Per Haplotype Reference Panel Update)",
        x = 0.5,
        y = 0,
        yref = "paper",
        xref = "paper",
        xanchor = "center",
        yanchor = "bottom",
        yshift = -35,
        showarrow = FALSE,
        font = list(size = 15) 
    ) 
fig %>% layout(yaxis = list(title = "Log-Likelihood"))
```
High Level of Drift, Fixed Number of Haplotypes in Reference Panel (x100)
```{r}
nhaps               = c(800) #Change the Nhaps in reference panel
nsnps               = c(50)#Constant NSnps
results             = c()                   #Store results as a vector
gwas_variants_af    = list()        #Needs to be a list 
Fst                 = 1e-1
nSamples            = 5      #Constant nSamples
alpha               = c(0.1,1,100,1e3)    #Constant alpha
weights_resolution  = 200
for (i in 1:length(alpha)) {
  chains                = 1       #Single Chain
  sim_data              = simulate_genetic_drift(Fst   = 0.01, 
                                                 nhaps = nhaps,
                                                 nsnps = nsnps, 
                                                 noise = alpha,
                                                 weights_resolution = weights_resolution)
  observed_gwas_se      = sim_data[[1]]
  gwas_variants_af[[i]] = sim_data[[2]]  #Store the GWAS variant AF as matrix 
  ref_hap_panel         = sim_data[[3]]         
  results[i]            = LD_from_GSHMM(ref_panel_haplotypes = ref_hap_panel,
                                         Fst                  = Fst,
                                         betas                = FALSE,
                                         alpha                = alpha,
                                         nSamples             = nSamples,
                                         recomb_rate          = 1e-100,
                                         weights_resolution   = weights_resolution,
                                         likelihood_toggle    = TRUE,
                                         se_observed          = observed_gwas_se,
                                         LD_Infer             = FALSE,
                                         genetic_map          = FALSE,
                                         chain_likelihood     = TRUE,
                                         nChains              = chains,
                                         recombination        = FALSE)
}

#Add on the traces across the different parameters (ie nHaps)
p <- plot_ly()
col     = brewer.pal(n = length(nhaps), name = "Dark2")
count = 1
r2    = list()
for (rep_number in 1:nreplicates) {
  for (i in 1:length(nhaps)) {
  ll = replicate_result[[rep_number]][i][[1]][[3]]
  nSamples_trace = seq(1:length(ll))/nhaps[i]
  r2[[count]] = r2_true_inferred_af_across_chains(inferred_allele_freq = replicate_result[[rep_number]][i][[1]][[4]],
                                         true_allele_freq     = gwas_variants_af[[count]],
                                         graphic              = FALSE,
                                         nSamples             = nSamples,
                                         nhaps                = nhaps[i])
  if (length(nSamples_trace) != length(r2)) {
    
  }

  if (rep_number == 1 ) {
     p <- add_trace(p, x=nSamples_trace, y=r2[[count]], type="scatter", mode="lines",evaluate = TRUE, 
                    name = sprintf("%s Ref Haplotypes",nhaps[i]),color=col[i], showlegend = T, opacity = 0.2)
    }
  else{
    p <- add_trace(p, x=nSamples_trace, y=r2[[count]], type="scatter", mode="lines",evaluate = TRUE, 
                    name = sprintf("%s Ref Haplotypes",nhaps[i]),color=col[i], showlegend = F, opacity = 0.2)
    }
    count = count + 1
  }
}

#Average across the r2's
#Split list into the groups
groups = list()
for (i in 1:length(nhaps)) {
  groups[[i]] = r2[seq(i,40,4)]
}

#Average 
r2_avg = list()
for (i in 1:length(groups)) {
    reshape = length(replicate_result[[1]][i][[1]][[3]])
    avg = rowMeans(matrix(unlist(groups[[i]]),reshape))
    r2_avg[[i]] = avg
    nSamples_trace = length(unlist(r2_avg[[i]]))
}

for (idx in 1:length(nhaps)) {
      ll = replicate_result[[rep_number]][idx][[1]][[3]]
      
      nSamples_trace = seq(1:length(ll))/nhaps[idx]
      p <- add_trace(p, x= nSamples_trace, y=unlist(r2_avg[[idx]]), 
                    type="scatter", mode="lines",evaluate = TRUE, 
                    showlegend = F, opacity = 1,color = col[idx])
}

p <- p %>% layout(title = 'Correlation of True to Inferred Allele Frequency Across Sweeps',
                    xaxis = list(title = 'nSamples (Full Sweeps)'),
                    yaxis = list(title = 'r2 correlation (Inferred~True)'))
p

ll_plot   = plot_ly(showlegend = TRUE)
ll_zoomed = plot_ly(showlegend = FALSE)
max_ll  = c()
col     = brewer.pal(n = length(nsnps), name = "Dark2")
for (i in 1:length(nsnps)) {
  ll = results[i][[1]][[3]]
  max_ll[i]  = max(ll)
  ll_plot <- add_trace(ll_plot,
                       x=seq(1:length(ll))/nhaps, 
                       y=ll, type="scatter", 
                       mode="markers",
                       evaluate = TRUE, name = sprintf("%s SNPS",nsnps[i]),
                       marker   = list(color = col[i]))
  ll_zoomed <- add_trace(ll_zoomed,
                       x=seq(1:length(ll))/nhaps, 
                       y=ll, type="scatter", 
                       mode="markers",
                       evaluate = TRUE, name = sprintf("%s SNPS",nsnps[i]),
                       marker   = list(color = col[i]))
}
ll_plot   = ll_plot %>% layout(title = 'Log Likelihood Across per Haplotype Sweeps',
                    xaxis = list(title = 'nSamples (Per Haplotype Sweeps)'),
                    yaxis = list(title = 'Log Likelihood'))
ll_zoomed = ll_zoomed  %>% layout(title = 'Log Likelihood Across per Haplotype Sweeps',
                    xaxis = list(title = 'nSamples (Per Haplotype Sweeps)'),
                    yaxis = list(range = c(0,max(max_ll)*1.1),
                                 title = 'Log Likelihood'))
ll_zoomed <- ll_zoomed %>% layout(marker     = list(color = col))

fig <- subplot(ll_plot,
               ll_zoomed) %>%  add_annotations(
        text = "Number of Samples (Per Haplotype Reference Panel Update)",
        x = 0.5,
        y = 0,
        yref = "paper",
        xref = "paper",
        xanchor = "center",
        yanchor = "bottom",
        yshift = -35,
        showarrow = FALSE,
        font = list(size = 15) 
    ) 
fig %>% layout(yaxis = list(title = "Log-Likelihood"))
```
How important is it to get the right level of drift in the inference algorithim?
```{r}
nhaps               = c(800) #Change the Nhaps in reference panel
nsnps               = c(50)                  #Constant NSnps
results             = c()                   #Store results as a vector
gwas_variants_af    = list()        #Needs to be a list 
Fst                 = c(0.001,0.01,0.1)
nSamples            = 5      #Constant nSamples
alpha               = 1e3     #Constant alpha
weights_resolution  = 200
for (i in 1:length(Fst)) {
  chains                = 1       #Single Chain
  sim_data              = simulate_genetic_drift(Fst   = 0.01, 
                                                 nhaps = nhaps,
                                                 nsnps = nsnps, 
                                                 noise = alpha,
                                                 weights_resolution = weights_resolution)
  observed_gwas_se      = sim_data[[1]]
  gwas_variants_af[[i]] = sim_data[[2]]  #Store the GWAS variant AF as matrix 
  ref_hap_panel         = sim_data[[3]]         
  results[i]            = LD_from_GSHMM(ref_panel_haplotypes = ref_hap_panel,
                                         Fst                  = Fst[i],
                                         betas                = FALSE,
                                         alpha                = alpha,
                                         nSamples             = nSamples,
                                         recomb_rate          = 1e-100,
                                         weights_resolution   = weights_resolution,
                                         likelihood_toggle    = TRUE,
                                         se_observed          = observed_gwas_se,
                                         LD_Infer             = FALSE,
                                         genetic_map          = FALSE,
                                         chain_likelihood     = TRUE,
                                         nChains              = chains,
                                         recombination        = FALSE)
}

#Add on the traces across the different parameters (ie nHaps)
p <- plot_ly()
for (i in 1:length(Fst)) {
  ll = results[i][[1]][[3]]
  nSamples_trace = seq(1:length(ll))/nhaps
  r2 = r2_true_inferred_af_across_chains(inferred_allele_freq = results[i][[1]][[4]],
                                         true_allele_freq     = gwas_variants_af[[i]],
                                         graphic              = FALSE,
                                         nSamples             = nSamples,
                                         nhaps                = nhaps)
  p <- add_trace(p, x=nSamples_trace, y=r2, type="scatter", mode="markers",evaluate = TRUE, name = sprintf("%s Fst",Fst[i]))
}
p <- p %>% layout(title = 'Correlation of True to Inferred Allele Frequency Across Sweeps',
                    xaxis = list(title = 'nSamples (Full Sweeps)'),
                    yaxis = list(title = 'r2 correlation (Inferred~True)'))
p

ll_plot   = plot_ly(showlegend = TRUE)
ll_zoomed = plot_ly(showlegend = FALSE)
max_ll  = c()
col     = brewer.pal(n = length(nsnps), name = "Dark2")
for (i in 1:length(nsnps)) {
  ll = results[i][[1]][[3]]
  max_ll[i]  = max(ll)
  ll_plot <- add_trace(ll_plot,
                       x=seq(1:length(ll))/nhaps, 
                       y=ll, type="scatter", 
                       mode="markers",
                       evaluate = TRUE, name = sprintf("%s SNPS",nsnps[i]),
                       marker   = list(color = col[i]))
  ll_zoomed <- add_trace(ll_zoomed,
                       x=seq(1:length(ll))/nhaps, 
                       y=ll, type="scatter", 
                       mode="markers",
                       evaluate = TRUE, name = sprintf("%s SNPS",nsnps[i]),
                       marker   = list(color = col[i]))
}
ll_plot   = ll_plot %>% layout(title = 'Log Likelihood Across per Haplotype Sweeps',
                    xaxis = list(title = 'nSamples (Per Haplotype Sweeps)'),
                    yaxis = list(title = 'Log Likelihood'))
ll_zoomed = ll_zoomed  %>% layout(title = 'Log Likelihood Across per Haplotype Sweeps',
                    xaxis = list(title = 'nSamples (Per Haplotype Sweeps)'),
                    yaxis = list(range = c(0,max(max_ll)*1.1),
                                 title = 'Log Likelihood'))
ll_zoomed <- ll_zoomed %>% layout(marker     = list(color = col))

fig <- subplot(ll_plot,
               ll_zoomed) %>%  add_annotations(
        text = "Number of Samples (Per Haplotype Reference Panel Update)",
        x = 0.5,
        y = 0,
        yref = "paper",
        xref = "paper",
        xanchor = "center",
        yanchor = "bottom",
        yshift = -35,
        showarrow = FALSE,
        font = list(size = 15) 
    ) 
fig %>% layout(yaxis = list(title = "Log-Likelihood"))
```
How does the number of states effect things?
```{r}
nhaps               = c(800) #Change the Nhaps in reference panel
nsnps               = c(100)                  #Constant NSnps
results             = c()                   #Store results as a vector
gwas_variants_af    = list()        #Needs to be a list 
Fst                 = 0.1
nSamples            = 5      #Constant nSamples
alpha               = 1e3     #Constant alpha
weights_resolution  = c(5,200)
for (i in 1:length(weights_resolution)) {
  chains                = 1       #Single Chain
  sim_data              = simulate_genetic_drift(Fst   = 0.1, 
                                                 nhaps = nhaps,
                                                 nsnps = nsnps, 
                                                 noise = alpha,
                                                 weights_resolution = 200)
  observed_gwas_se      = sim_data[[1]]
  gwas_variants_af[[i]] = sim_data[[2]]  #Store the GWAS variant AF as matrix 
  ref_hap_panel         = sim_data[[3]]         
  results[i]            = LD_from_GSHMM(ref_panel_haplotypes = ref_hap_panel,
                                         Fst                  = Fst,
                                         betas                = FALSE,
                                         alpha                = alpha,
                                         nSamples             = nSamples,
                                         recomb_rate          = 1e-100,
                                         weights_resolution   = weights_resolution[i],
                                         likelihood_toggle    = TRUE,
                                         se_observed          = observed_gwas_se,
                                         LD_Infer             = FALSE,
                                         genetic_map          = FALSE,
                                         chain_likelihood     = TRUE,
                                         nChains              = chains,
                                         recombination        = FALSE)
}

#Add on the traces across the different parameters (ie nHaps)
p <- plot_ly()
for (i in 1:length(weights_resolution)) {
  ll = results[i][[1]][[3]]
  nSamples_trace = seq(1:length(ll))/nhaps
  r2 = r2_true_inferred_af_across_chains(inferred_allele_freq = results[i][[1]][[4]],
                                         true_allele_freq     = gwas_variants_af[[i]],
                                         graphic              = FALSE,
                                         nSamples             = nSamples,
                                         nhaps                = nhaps)
  p <- add_trace(p, x=nSamples_trace, y=r2, type="scatter", mode="markers",evaluate = TRUE, name = sprintf("%s Weights",weights_resolution[i]))
}
p <- p %>% layout(title = 'Correlation of True to Inferred Allele Frequency Across Sweeps',
                    xaxis = list(title = 'nSamples (Full Sweeps)'),
                    yaxis = list(title = 'r2 correlation (Inferred~True)'))
p

ll_plot   = plot_ly(showlegend = TRUE)
ll_zoomed = plot_ly(showlegend = FALSE)
max_ll  = c()
col     = brewer.pal(n = length(nsnps), name = "Dark2")
for (i in 1:length(nsnps)) {
  ll = results[i][[1]][[3]]
  max_ll[i]  = max(ll)
  ll_plot <- add_trace(ll_plot,
                       x=seq(1:length(ll))/nhaps, 
                       y=ll, type="scatter", 
                       mode="markers",
                       evaluate = TRUE, name = sprintf("%s SNPS",nsnps[i]),
                       marker   = list(color = col[i]))
  ll_zoomed <- add_trace(ll_zoomed,
                       x=seq(1:length(ll))/nhaps, 
                       y=ll, type="scatter", 
                       mode="markers",
                       evaluate = TRUE, name = sprintf("%s SNPS",nsnps[i]),
                       marker   = list(color = col[i]))
}
ll_plot   = ll_plot %>% layout(title = 'Log Likelihood Across per Haplotype Sweeps',
                    xaxis = list(title = 'nSamples (Per Haplotype Sweeps)'),
                    yaxis = list(title = 'Log Likelihood'))
ll_zoomed = ll_zoomed  %>% layout(title = 'Log Likelihood Across per Haplotype Sweeps',
                    xaxis = list(title = 'nSamples (Per Haplotype Sweeps)'),
                    yaxis = list(range = c(0,max(max_ll)*1.1),
                                 title = 'Log Likelihood'))
ll_zoomed <- ll_zoomed %>% layout(marker     = list(color = col))

fig <- subplot(ll_plot,
               ll_zoomed) %>%  add_annotations(
        text = "Number of Samples (Per Haplotype Reference Panel Update)",
        x = 0.5,
        y = 0,
        yref = "paper",
        xref = "paper",
        xanchor = "center",
        yanchor = "bottom",
        yshift = -35,
        showarrow = FALSE,
        font = list(size = 15) 
    ) 
fig %>% layout(yaxis = list(title = "Log-Likelihood"))
```
Can probably just stick to 5 weight states(?)
```{r}
#How well do we do with varying the noise?
nhaps               = c(800) #Change the Nhaps in reference panel
nsnps               = c(50)                  #Constant NSnps
results             = c()                   #Store results as a vector
gwas_variants_af    = list()        #Needs to be a list 
Fst                 = 0.1
nSamples            = 5      #Constant nSamples
alpha               = c(1,10,500,1000)    #Constant alpha
weights_resolution  = 5
for (i in 1:length(alpha)) {
  chains                = 1       #Single Chain
  sim_data              = simulate_genetic_drift(Fst   = 0.005, 
                                                 nhaps = nhaps,
                                                 nsnps = nsnps, 
                                                 noise = 1,
                                                 weights_resolution = 200)
  observed_gwas_se      = sim_data[[1]]
  gwas_variants_af[[i]] = sim_data[[2]]  #Store the GWAS variant AF as matrix 
  ref_hap_panel         = sim_data[[3]]         
  results[i]            = LD_from_GSHMM(ref_panel_haplotypes = ref_hap_panel,
                                         Fst                  = Fst,
                                         betas                = FALSE,
                                         alpha                = alpha[i],
                                         nSamples             = nSamples,
                                         recomb_rate          = 1e-100,
                                         weights_resolution   = weights_resolution,
                                         likelihood_toggle    = TRUE,
                                         se_observed          = observed_gwas_se,
                                         LD_Infer             = FALSE,
                                         genetic_map          = FALSE,
                                         chain_likelihood     = TRUE,
                                         nChains              = chains,
                                         recombination        = FALSE)
}

#Add on the traces across the different parameters (ie nHaps)
p <- plot_ly()
for (i in 1:length(alpha)) {
  ll = results[i][[1]][[3]]
  nSamples_trace = seq(1:length(ll))/nhaps
  r2 = r2_true_inferred_af_across_chains(inferred_allele_freq = results[i][[1]][[4]],
                                         true_allele_freq     = gwas_variants_af[[i]],
                                         graphic              = FALSE,
                                         nSamples             = nSamples,
                                         nhaps                = nhaps)
  p <- add_trace(p, x=nSamples_trace, y=r2, type="scatter", mode="markers",evaluate = TRUE, name = sprintf("%s Noise",alpha[i]))
}
p <- p %>% layout(title = 'Correlation of True to Inferred Allele Frequency Across Sweeps',
                    xaxis = list(title = 'nSamples (Full Sweeps)'),
                    yaxis = list(title = 'r2 correlation (Inferred~True)'))
p

ll_plot   = plot_ly(showlegend = TRUE)
ll_zoomed = plot_ly(showlegend = FALSE)
max_ll  = c()
col     = brewer.pal(n = length(nsnps), name = "Dark2")
for (i in 1:length(nsnps)) {
  ll = results[i][[1]][[3]]
  max_ll[i]  = max(ll)
  ll_plot <- add_trace(ll_plot,
                       x=seq(1:length(ll))/nhaps, 
                       y=ll, type="scatter", 
                       mode="markers",
                       evaluate = TRUE, name = sprintf("%s SNPS",nsnps[i]),
                       marker   = list(color = col[i]))
  ll_zoomed <- add_trace(ll_zoomed,
                       x=seq(1:length(ll))/nhaps, 
                       y=ll, type="scatter", 
                       mode="markers",
                       evaluate = TRUE, name = sprintf("%s SNPS",nsnps[i]),
                       marker   = list(color = col[i]))
}
ll_plot   = ll_plot %>% layout(title = 'Log Likelihood Across per Haplotype Sweeps',
                    xaxis = list(title = 'nSamples (Per Haplotype Sweeps)'),
                    yaxis = list(title = 'Log Likelihood'))
ll_zoomed = ll_zoomed  %>% layout(title = 'Log Likelihood Across per Haplotype Sweeps',
                    xaxis = list(title = 'nSamples (Per Haplotype Sweeps)'),
                    yaxis = list(range = c(0,max(max_ll)*1.1),
                                 title = 'Log Likelihood'))
ll_zoomed <- ll_zoomed %>% layout(marker     = list(color = col))

fig <- subplot(ll_plot,
               ll_zoomed) %>%  add_annotations(
        text = "Number of Samples (Per Haplotype Reference Panel Update)",
        x = 0.5,
        y = 0,
        yref = "paper",
        xref = "paper",
        xanchor = "center",
        yanchor = "bottom",
        yshift = -35,
        showarrow = FALSE,
        font = list(size = 15) 
    ) 
fig %>% layout(yaxis = list(title = "Log-Likelihood"))
```

